Framework/Crystal/src/PredictPeaks.cpp

```diff
@@ -461,37 +470,86 @@ void PredictPeaks::calculateQAndAddToOutput(const V3D &hkl,
-  if (!p.getDetector() &&
-      !(useExtendedDetectorSpace &&
-        m_inst->getComponentByName("extended-detector-space")))
+  const V3D q = orientedUB * hkl * (2.0 * M_PI * m_qConventionFactor);
+
+  const auto convention =
+      Kernel::ConfigService::Instance().getString("Q.convention");
+  double norm_q = q.norm();
+  boost::shared_ptr<const ReferenceFrame> refFrame =
+      this->m_inst->getReferenceFrame();
+  const V3D refBeamDir = refFrame->vecPointingAlongBeam();
+  // Default for ki-kf has -q
+  const auto qSign = (convention == "Crystallography") ? -1.0 : 1.0;
+  const double qBeam = q.scalar_prod(refBeamDir) * qSign;
+  double one_over_wl = (norm_q * norm_q) / (2.0 * qBeam);
+  double wl = (2.0 * M_PI) / one_over_wl;
+  // Default for ki-kf has -q
+  const auto inverseQSign = (convention == "Crystallography") ? 1.0 : -1.0;
+  V3D detectorDir = q * inverseQSign;
+  detectorDir[refFrame->pointingAlongBeam()] = one_over_wl - qBeam;
+  detectorDir.normalize();
+
+  const bool useExtendedDetectorSpace =
+      getProperty("PredictPeaksOutsideDetectors");
+  const auto result = m_detectorCacheSearch->findDetectorIndex(q);
+  const auto hitDetector = std::get<0>(result);
+  const auto index = std::get<1>(result);
+
+  if (!hitDetector && !useExtendedDetectorSpace) {
     return;
+  }
 
-  // Only add peaks that hit the detector
-  p.setGoniometerMatrix(goniometerMatrix);
-  // Save the run number found before.
-  p.setRunNumber(m_runNumber);
-  p.setHKL(hkl * m_qConventionFactor);
+  const auto &detInfo = m_pw->detectorInfo();
+  const auto &det = detInfo.detector(index);
+  if (hitDetector) {
+    // peak hit a detector to add it to the list
+    Peak peak(m_inst, det.getID(), wl);
+    if (!peak.getDetector())
+      return;
+
+    // Only add peaks that hit the detector
+    peak.setGoniometerMatrix(goniometerMatrix);
+    // Save the run number found before.
+    peak.setRunNumber(m_runNumber);
+    peak.setHKL(hkl * m_qConventionFactor);
+
+    if (m_sfCalculator) {
+      peak.setIntensity(m_sfCalculator->getFSquared(hkl));
+    }
 
-  if (m_sfCalculator) {
-    p.setIntensity(m_sfCalculator->getFSquared(hkl));
-  }
+    // Add it to the workspace
+    m_pw->addPeak(peak);
+  } else if (useExtendedDetectorSpace) {
+    // use extended detector space to try and guess peak position
+    const auto component =
+        m_inst->getComponentByName("extended-detector-space");
+    const auto c = boost::dynamic_pointer_cast<const ObjComponent>(component);
+    if (!c)
+      throw std::runtime_error("PredictPeaks: user requested use of a extended "
+                               "detector space to predict peaks but there is no"
+                               "definition in the IDF");
+
+    // find where this Q vector should intersect with "extended" space
+    Geometry::Track track(detInfo.samplePosition(), detectorDir);
+    if (!c->interceptSurface(track))
+      return;
+
+    const auto magnitude = track.back().exitPoint.norm();
```

Comment:
As I understand it when you trace a ray to the surface of some object, if the object was hit you will get a vector from the starting point to the exit point (where it hit). I believe it is possible to "chain" rays more than once, but in "interceptSurface" it only adds one vector hence I use `back()` here.

Perhaps I should add a comment in the code to clarify as well.